pkg/volume/util/nestedpendingoperations/nestedpendingoperations_test.go

```diff
@@ -759,6 +759,151 @@ func Test_NestedPendingOperations_Positive_Issue_88355(t *testing.T) {
+	volumeName := v1.UniqueVolumeName("test-volume")
+	podName1 := volumetypes.UniquePodName("pod1")
+	podName2 := volumetypes.UniquePodName("pod2")
+	podName3 := volumetypes.UniquePodName("pod3")
+	podName4 := EmptyUniquePodName
+	nodeName := EmptyNodeName
+
+	// delay after an operation is signaled to finish to ensure it actually
+	// finishes before running the next operation.
+	delay := 50 * time.Millisecond
+	backoffDelay := 500 * time.Millisecond
+
+	// fake operation1 for pod1 fails
+	operation1DoneCh := make(chan interface{})
+	operation1 := generateWaitWithErrorFunc(operation1DoneCh)
+	err1 := grm.Run(volumeName, podName1, nodeName /* nodeName */, volumetypes.GeneratedOperations{OperationFunc: operation1, OperationName: "umount"})
+	if err1 != nil {
+		t.Errorf("NestedPendingOperations failed. Expected: <no error> Actual: <%v>", err1)
+	}
+
+	// fake operation2 for pod2 fails
+	operation2DoneCh := make(chan interface{})
+	operation2 := generateWaitWithErrorFunc(operation2DoneCh)
+	err2 := grm.Run(volumeName, podName2, nodeName /* nodeName */, volumetypes.GeneratedOperations{OperationFunc: operation2, OperationName: "umount"})
+	if err2 != nil {
+		t.Errorf("NestedPendingOperations failed. Expected: <no error> Actual: <%v>", err2)
+	}
+
+	// fake operation3 for pod3 fails
+	operation3DoneCh := make(chan interface{})
+	operation3 := generateWaitWithErrorFunc(operation3DoneCh)
+	err3 := grm.Run(volumeName, podName3, nodeName /* nodeName */, volumetypes.GeneratedOperations{OperationFunc: operation3, OperationName: "umount"})
+	if err3 != nil {
+		t.Errorf("NestedPendingOperations failed. Expected: <no error> Actual: <%v>", err3)
+	}
+
+	operation1DoneCh <- true
+	operation2DoneCh <- true
+	operation3DoneCh <- true
+	time.Sleep(delay)
+
+	// fake operation4 with EmptyUniquePodName fails
+	operation4DoneCh := make(chan interface{})
+	operation4 := generateWaitWithErrorFunc(operation4DoneCh)
+	err4 := grm.Run(volumeName, podName4, nodeName /* nodeName */, volumetypes.GeneratedOperations{OperationFunc: operation4, OperationName: "mount"})
+	if err4 != nil {
+		t.Errorf("NestedPendingOperations failed. Expected: <no error> Actual: <%v>", err4)
+	}
+
+	operation4DoneCh <- true
+
+	// operation for pod2 retry
+	time.Sleep(backoffDelay)
+	operation5 := noopFunc
+	err5 := grm.Run(volumeName, podName2, nodeName, volumetypes.GeneratedOperations{OperationFunc: operation5, OperationName: "umount"})
+	if err5 != nil {
+		t.Errorf("NestedPendingOperations failed. Expected: <no error> Actual: <%v>", err5)
+	}
+	time.Sleep(delay)
+
+	// Assert
+	// operation4 override operation1 or operation3, and operation5 will override operation2,
+	// so finally, only operation4, operation1 or operation3 left
+	grm.(*nestedPendingOperations).lock.Lock()
+	defer grm.(*nestedPendingOperations).lock.Unlock()
+	l := len(grm.(*nestedPendingOperations).operations)
+	if l != 2 {
+		t.Errorf("NestedPendingOperations length should return 2, got: %v", l)
+	}
+}
```

Comment:
I don't think these tests exercise the bug under question correctly. They are passing without the changes into nestedpendingoperations file.

Reply:
@gnufied Actually ` grm.Run` would call isOperationExists to do the operations, so I think these tests cover these modifications

Reply:
I am not sure. I tried running the tests without incorporating any of the changes from nestedpending_operations file (that is the actual code change) and the test still pass. Am i missing something? Will tests fail in unpredictable manner without the change but do not reproduce the bug 100% of the time?

Reply:
@gnufied okay, i see, will fix these tests in next push